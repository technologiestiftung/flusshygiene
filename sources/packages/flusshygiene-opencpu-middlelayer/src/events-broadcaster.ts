import { EventEmitter } from 'events';

declare interface BroadCaster {
  on(event: 'passthrough', listener: (name: any) => void): this;
  on(event: string, listener: Function): this;
}

class BroadCaster extends EventEmitter {
  public static getInstance() {
    if (!BroadCaster.instance) {
      BroadCaster.instance = new BroadCaster('Singleton');
    }
    return BroadCaster.instance;
  }
  private static instance: BroadCaster;
  private constructor(public readonly name: string) {
    super();
  }
}

export { BroadCaster };
