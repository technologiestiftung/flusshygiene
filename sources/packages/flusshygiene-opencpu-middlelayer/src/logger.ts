import * as winston from 'winston';

const logger = winston.createLogger({
  exitOnError: false,
  level: 'info',
  transports: [
    new winston.transports.Console({
      level: 'debug',
      handleExceptions: true,
    }),
    new winston.transports.File({
      filename: 'logs/pubsub-all.log',
      handleExceptions: true,
      maxsize: 5242880, //5MB
      maxFiles: 5,
    }),
  ],
});

const logStream = {
  write: (text: string) => {
    logger.info(text);
  },
};

export { logStream, logger };
