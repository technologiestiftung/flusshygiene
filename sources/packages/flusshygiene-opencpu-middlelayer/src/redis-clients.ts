// import redis, { ClientOpts } from 'redis';
// const opts: ClientOpts = {
//   host: process.env.REDIS_HOST || '127.0.0.1',
//   port:
//     process.env.REDIS_PORT === undefined
//       ? 6379
//       : parseInt(process.env.REDIS_PORT, 10),
//   retry_strategy: () => 1000,
// };
// const redisClient = redis.createClient(opts);
// const sub = redisClient.duplicate();

// export const pubsub = async () => {
//   try {
//     sub.on('message', (channel, message) => {
//       console.log(`Redis Channel ${channel} => ${message}\n`);
//     });
//     sub.on('connect', () => {
//       console.log('sub is connected');
//     });
//     sub.subscribe('hello');
//   } catch (err) {
//     throw err;
//   }
// };
