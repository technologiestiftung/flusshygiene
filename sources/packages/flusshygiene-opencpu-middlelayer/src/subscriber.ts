import redis, { ClientOpts } from 'redis';
const opts: ClientOpts = {
  host: process.env.REDIS_HOST || 'redis',
  port:
    process.env.REDIS_PORT === undefined
      ? 6379
      : parseInt(process.env.REDIS_PORT, 10),
  retry_strategy: () => 1000,
};
const redisClient = redis.createClient(opts);
const sub = redisClient.duplicate();

// export const pubsub = async () => {
// try {
sub.on('message', (channel, message) => {
  console.log(`Redis Channel ${channel} => ${message}\n`);
});
sub.subscribe('hello');
// } catch (err) {
// throw err;
// }
// };
