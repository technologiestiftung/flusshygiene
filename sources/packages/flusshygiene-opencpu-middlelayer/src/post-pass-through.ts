import { IObject } from './common/interfaces';
import got from 'got';
import { logger } from './logger';

/**
 * Pass POST Request through to another server
 */
const postPassThrough: (
  url: string,
  body: IObject,
) => Promise<IObject> = async (url, body) => {
  try {
    const gres = await got.post(url, {
      json: true,
      body,
    });
    return gres.body;
  } catch (error) {
    if (process.env.NODE_ENV === 'development') {
      console.error(error);
    } else {
      logger.error(error);
    }
    return error;
  }
};

export { postPassThrough };
