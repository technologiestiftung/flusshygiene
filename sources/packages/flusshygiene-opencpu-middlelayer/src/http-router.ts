import { OCPU_API_HOST, VERSION } from './common/constants';

import { BroadCaster } from './events-broadcaster';
import Router from 'express-promise-router';
import { postPassThrough } from './post-pass-through';

enum FHpredictFunctions {
  calibrate = 'provide_rain_data_for_bathing_spot',
  model = 'simple',
  predict = 'simple',
}

// import got from 'got';
const router = Router();
const broadcaster = BroadCaster.getInstance();

router.get('/health', async (_req, res) => {
  // req.session!.name = 'foo';
  // console.log(req.session);
  res.json({
    success: true,
    message: 'Everything OK',
    version: VERSION,
  });
});

router.post(['/calibrate', '/predict', '/model'], async (req, res) => {
  console.log(req.sessionID);
  // if (req.body.payload === undefined) {
  //   res.status(400).json({
  //     success: false,
  //     message:
  //       'You need to pass the "payload" [Object] in the body of your POST request. The payload will be the body this module passes through',
  //     version: VERSION,
  //     sessionID: req.sessionID,
  //   });
  //   return;
  // }
  let url = '';
  switch (req.url) {
    case '/calibrate':
      url = `${OCPU_API_HOST}/ocpu/library/fhpredict/R/${FHpredictFunctions.calibrate}/json`;
      break;
    case '/predict':
      url = `${OCPU_API_HOST}/ocpu/library/fhpredict/R/${FHpredictFunctions.predict}/json`;
      break;
    case '/model':
      url = `${OCPU_API_HOST}/ocpu/library/fhpredict/R/${FHpredictFunctions.model}/json`;
      break;
  }

  res.status(201).json({
    success: true,
    message: `Your request to ${req.url} is beeing processed`,
    version: VERSION,
    sessionID: req.sessionID,
  });
  broadcaster.emit('passthrough', 'start');
  // const passThroughBody = { ...req.body };
  postPassThrough(url, req.body)
    .then((body) => {
      // console.log(body);
      broadcaster.emit('passthrough', { body: body, sessionID: req.sessionID });
      broadcaster.emit('passthrough', 'end');
    })
    .catch((error) => {
      broadcaster.emit('passthrough', error);
      broadcaster.emit('passthrough', 'end');
      // console.error(error);
      // throw error;
    });

  // const gres = await got.post('http://localhost:4444/process/300000', {
  //   json: true,
  //   body: passThroughBody,
  // });
  // console.log('response body', gres.body);
});
// router.post('/predict', async (req, res) => {
//   res.status(201).json({
//     success: true,
//     message: `Your request to ${req.url} is beeing processed`,
//   });
// });
// router.post('/model', async (req, res) => {
//   res.status(201).json({
//     success: true,
//     message: `Your request to ${req.url} is beeing processed`,
//   });
// });

export { router };
