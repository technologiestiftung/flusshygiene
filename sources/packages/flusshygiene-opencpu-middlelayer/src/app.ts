import { SESSION_SECRET } from './common/constants';
import cors from 'cors';
import express from 'express';
import helmet from 'helmet';
import { logStream } from './logger';
import morgan from 'morgan';
import { router } from './http-router';
import session from 'express-session';

const app = express();
app.use(
  session({
    secret:
      SESSION_SECRET !== undefined
        ? SESSION_SECRET
        : '45364f84-f39c-449d-b997-27f20da7e8ec',
    resave: false,
    saveUninitialized: true,
  }),
);
app.use(cors());
if (process.env.NODE_ENV === 'development') {
  app.use(morgan('combined'));
} else if (process.env.NODE_ENV === 'test') {
} else {
  app.use(helmet());
  app.use(morgan('combined', { stream: logStream }));
}
app.use(express.json({ limit: '10mb' }));
app.use(express.urlencoded({ extended: true, limit: '10mb' }));
app.use('/', router);
export default app;
