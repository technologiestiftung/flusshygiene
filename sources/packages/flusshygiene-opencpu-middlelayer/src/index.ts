import { websocketServer, wsSubmit } from './websocket-server';

import { BroadCaster } from './events-broadcaster';
import { ENV_SUFFIX } from './common/constants';
import app from './app';
import http from 'http';
import { logger } from './logger';

const broadcaster = BroadCaster.getInstance();
// const ENV_SUFFIX = process.env.NODE_ENV === 'production' ? 'PROD' : 'DEV';

const PORT: number | string =
  process.env[`REDIS_EXPRESS_PORT_${ENV_SUFFIX}`] || 4004;

const server = http.createServer(app);
const wss = websocketServer(server);

broadcaster.on('passthrough', (data: any) => {
  switch (data) {
    case 'start': {
      console.info('passthrough has started');

      break;
    }
    case 'end': {
      console.info('passthrough has ended');
      break;
    }
    default: {
      // console.info('data', data);
      wsSubmit(wss, data);
      break;
    }
  }
});
server.listen(PORT, () => {
  if (process.env.NODE_ENV === 'development') {
    logger.info(
      `server listening on http://localhost:${PORT}\n`,
      `wss listening on ws://localhost:${PORT}`,
    );
  }
});

// pubsub().catch((err) => {
//   throw err;
// });
