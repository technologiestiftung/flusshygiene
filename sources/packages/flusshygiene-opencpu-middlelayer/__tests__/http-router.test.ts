import redis from 'redis';
import redis_mock from 'redis-mock';
const mockedRedis = jest
  .spyOn(redis, 'createClient')
  .mockImplementation(redis_mock.createClient);

import * as passThrough from '../src/post-pass-through';

import { BroadCaster } from '../src/events-broadcaster';
import app from '../src/app';
import nock from 'nock';
import request from 'supertest';

jest.useFakeTimers();
const scope = nock('http://localhost:4444')
  .post('/middlelayer/calibrate')
  .reply(201, {})
  .post('/middlelayer/predict')
  .reply(201, {})
  .post('/middlelayer/model')
  .reply(201, {})
  .post('/middlelayer/foo/bah')
  .reply(201, {});
afterAll(() => {
  mockedRedis.mockRestore();
});

describe('basic route tests', () => {
  test('should responde with success on calibrate', async (done) => {
    const response = await request(app)
      .post('/middlelayer/calibrate')
      .set('Cookie', ['connect.sid=4e89f412-cad5-4787-8d98-20212336b2c5'])
      .send({ url: 'http://localhost:4444/foo/bah', payload: {} });
    expect(response.status).toBe(201);
    expect(response.body).toMatchSnapshot({
      sessionID: expect.any(String),
      version: expect.any(String),
    });
    done();
  });
  test('should responde with success on predict', async (done) => {
    const response = await request(app)
      .post('/middlelayer/predict')
      .send({ url: 'http://localhost:4444/foo/bah', payload: {} });
    expect(response.status).toBe(201);
    expect(response.body).toMatchSnapshot({
      sessionID: expect.any(String),
      version: expect.any(String),
    });
    done();
  });
  test('should responde with success on model', async (done) => {
    const response = await request(app)
      .post('/middlelayer/model')
      .send({ url: 'http://localhost:4444/foo/bah', payload: {} });
    expect(response.status).toBe(201);
    expect(response.body).toMatchSnapshot({
      sessionID: expect.any(String),
    });
    done();
  });

  test('should respond with 404 on non existing route', async (done) => {
    const response = await request(app)
      .post('/middlelayer/foo')
      .send({});
    // console.log(response);
    expect(response.status).toBe(404);
    expect(response.text).toMatchSnapshot();
    done();
  });

  test.todo(
    'should respond with 404 due to missing payload' /*, async (done) => {
    const response = await request(app)
      .post('/middlelayer/calibrate')
      .send({ url: 'http://localhost:4444/foo/bah' });
    expect(response.status).toBe(400);
    expect(response.body.success).toBe(false);
    done();
  }*/,
  );

  test('should call broadcaster after working request', async (done) => {
    const broadcaster = BroadCaster.getInstance();
    const mockBroadcaster = jest.spyOn(broadcaster, 'emit');
    await request(app)
      .post('/middlelayer/model')
      .send({ url: 'http://localhost:4444/foo/bah', payload: {} });
    expect(mockBroadcaster).toHaveBeenCalled();
    mockBroadcaster.mockRestore();
    done();
  });
  test('should call broadcaster after working request', async (done) => {
    const broadcaster = BroadCaster.getInstance();
    const mockBroadcasterEmit = jest.spyOn(broadcaster, 'emit');
    const mockPassthrough = jest.spyOn(passThrough, 'postPassThrough');
    await request(app)
      .post('/middlelayer/model')
      .send({ url: 'http://localhost:4444/foo/bah', payload: {} });
    expect(mockBroadcasterEmit).toHaveBeenCalled();
    expect(mockPassthrough).toHaveBeenCalled();
    // console.log(mockBroadcasterEmit.mock);
    // console.log(mockPassthrough.mock);
    expect(mockPassthrough.mock.results[0].value).resolves.toBe({});
    mockPassthrough.mockRestore();
    mockBroadcasterEmit.mockRestore();
    done();
  });
});
