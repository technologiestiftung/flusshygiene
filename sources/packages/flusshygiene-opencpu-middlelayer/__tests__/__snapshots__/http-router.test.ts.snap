// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic route tests should respond with 404 on non existing route 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
<head>
<meta charset=\\"utf-8\\">
<title>Error</title>
</head>
<body>
<pre>Cannot POST /middlelayer/foo</pre>
</body>
</html>
"
`;

exports[`basic route tests should responde with success on calibrate 1`] = `
Object {
  "message": "Your request to /calibrate is beeing processed",
  "sessionID": Any<String>,
  "success": true,
  "version": "0.2.0",
}
`;

exports[`basic route tests should responde with success on model 1`] = `
Object {
  "message": "Your request to /model is beeing processed",
  "sessionID": Any<String>,
  "success": true,
  "version": "0.2.0",
}
`;

exports[`basic route tests should responde with success on predict 1`] = `
Object {
  "message": "Your request to /predict is beeing processed",
  "sessionID": Any<String>,
  "success": true,
  "version": "0.2.0",
}
`;
